armi/physics/neutronics/globalFlux/globalFluxInterface.py

```diff
@@ -1313,6 +1313,82 @@ def computeDpaRate(mgFlux, dpaXs):
     return dpaPerSecond
 
 
+def calcReactionRatesBlockList(objList, keff, xsNucDict):
+    r"""
+    Compute 1-group reaction rates for the objcects in objList (usually a block).
+
+    Parameters
+    ----------
+    objList : List[Block]
+        The list of objects to compute reaction rates on. Notionally this could be upgraded to be
+        any kind of ArmiObject but with params defined as they are it currently is only
+        implemented for a block.
+
+    keff : float
+        The keff of the core. This is required to get the neutron production rate correct
+        via the neutron balance statement (since nuSigF has a 1/keff term).
+
+    xsNucDict: Dict[str, XSNuclide]
+        Microscopic cross sections to use in computing the reaction rates.
+
+    .. impl:: Return the reaction rates for a given ArmiObject
+        :id: I_ARMI_FLUX_RX_RATES_BY_XS_ID
+        :implements: R_ARMI_FLUX_RX_RATES
+
+        This is an alternative implementation of :need:`I_ARMI_FLUX_RX_RATES` that
+        is more efficient when computing reaction rates for a large set of blocks
+        that share a common set of microscopic cross sections.
+
+        Fore more detail on the reation rate calculations, see :need:`I_ARMI_FLUX_RX_RATES`.
+    """
+    for obj in objList:
+        rate = {}
+        for simple in RX_PARAM_NAMES:
+            rate[simple] = 0.0
+
+        numberDensities = obj.getNumberDensities()
+        try:
+            mgFlux = np.array(obj.getMgFlux())
+        except TypeError:
+            continue
+
+        for nucName, numberDensity in numberDensities.items():
+            if numberDensity == 0.0:
+                continue
+            nucrate = {}
+            for simple in RX_PARAM_NAMES:
+                nucrate[simple] = 0.0
+
+            micros = xsNucDict[nucName].micros
+
+            # absorption is fission + capture (no n2n here)
+            for name in RX_ABS_MICRO_LABELS:
+                volumetricRR = numberDensity * mgFlux.dot(micros[name])
+                nucrate["rateAbs"] += volumetricRR
+                if name != "fission":
+                    nucrate["rateCap"] += volumetricRR
+                else:
+                    nucrate["rateFis"] += volumetricRR
+                    # scale nu by keff.
+                    nusigmaF = micros["fission"] * micros.neutronsPerFission
+                    nucrate["rateProdFis"] += (
+                        numberDensity * mgFlux.dot(nusigmaF) / keff
+                    )
+
+            nucrate["rateProdN2n"] += 2.0 * numberDensity * mgFlux.dot(micros.n2n)
+
+            for simple in RX_PARAM_NAMES:
+                if nucrate[simple]:
+                    rate[simple] += nucrate[simple]
+
+        for paramName, val in rate.items():
+            obj.p[paramName] = val  # put in #/cm^3/s
+
+        vFuel = obj.getComponentAreaFrac(Flags.FUEL) if rate["rateFis"] > 0.0 else 1.0
```

Comment:
I understand that you copied this line from the `calcReactionRates()` function, but what the heck is going on here?

I assume `vFuel` is meant to mean "volume of fuel", but instead of calculating volume it gets the area fraction. In the case of axially-extruded blocks, _maybe_ ...sure...

But if the fission rate is 0 then we somehow want the volume of fuel to be 1.0? Huh? I do not understand what is happening here.

Reply:
I guess it is just so that we don't do a `0/0` on the next line?

It seems like this little chunk of code could be more clear.

Reply:
My guess is that somebody was just going for the minimum number of lines of code to get the algebraically correct result.

I could make it more clear with more lines. Something like this:

![image](https://github.com/user-attachments/assets/f2926719-ef9c-46a6-ab7c-268ab726f271)

Reply:
Looks like you already did it. Great!